elastic/kibana
# [User Experience] Add error boundary to prevent UX dashboard from crashing the application

## Summary

This PR fixes #117543 by adding an Error Boundary to the UX dashboard, which was crashing when a user changed filters on a Kibana instance pointing to a cluster configured for CCS.

**NOTE:** Please notice that we haven't yet fixed the underlying root cause of the crash, as explained in https://github.com/elastic/kibana/issues/116226. This PR includes _exclusively_ the error boundary to prevent the error from bubbling up.

### Testing & Other related changes

Besides adding the Error Boundary, this PR includes a test to confirm errors will be caught.

For that test to be written cleanly, however, I've had to include in this PR:

1. A small refactor to the code which provides the `core` mocks to components. Now, it uses the core testing utilities rather than repeating the manual setup inline. Therefore, I can reuse it in both of the root app tests.
2. A rename for one of the `core` mock methods which had spelt the word `parameters `as `paramters`.

When it comes to E2E (or "Functional" as we call it) tests, we decided we wouldn't add one because we should ultimately fix the underlying behaviour, and, therefore, an E2E test wouldn't be value. That's as per described in the issue.

### Further improvements

The most difficult part of this PR was figuring out how to appropriately mock the `props` passed to the root app. Mocking one of its inner components made the automated testing easier, but it was still somewhat confusing to mock all the core objects and understand which pieces of `context` were being pulled in different places or which settings were being read.

Going forward it could be interesting to streamline this mocking process and improve the testing utilities.

### Old Behaviour

![error-boundary-ux-screen](https://user-images.githubusercontent.com/6868147/140095623-d211aff2-1496-4e1e-846d-58783f92b5ba.gif)

### New Behaviour (with Error Boundary)

![error-boundary-added](https://user-images.githubusercontent.com/6868147/140393869-29fca78e-b8a8-459a-8b84-3112402f4764.gif)

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

(I have deleted this section as I don't think it applies to this PR, but please let me know if I misunderstood)


### For maintainers


* I've also pushed a commit as `lucas.fernandesdacosta@elastic.co` so that the automations could be run for my GitHub account going forward.
* As this is my very first code-contribution to Kibana, I'd appreciate feedback and guidance on whether I've used the correct labels and followed the process adequately.
* I've marked this PR as draft for now because I've seen problems related to the remote cluster I'm testing against, so I'll look into that before marking this PR as ready for review (`circuit_breaking_exception: [circuit_breaking_exception] Reason: [parent] Data too large, data for [indices:data/read/search] would be [2117526278/1.9gb]`)

### How to Verify this PR:

1. Use [`oblt-cli`](https://github.com/elastic/observability-test-environments/tree/master/tools/oblt_cli) to create a deployment configured for CCS against `edge-oblt`.
2. [Run Kibana from source (this PR)](https://www.elastic.co/guide/en/kibana/master/development-getting-started.html) and configure it to connect to your deployment, as per the instructions in the `edge-oblt` tool `README`.
3. Log into Kibana and access the UX dashboard.
4. Change a filter in the UX dashboard, like the location filter.
5. You should see that the error is caught by the error boundary and you can continue to use the app

## Release note

The User Experience Dashboard errors will now be captured and appropriately displayed, preventing the application from crashing.
